src/renderer/atlas/AtlasEngine.cpp

```diff
@@ -190,46 +124,115 @@ try
     _p.cursorRect = {};
     _p.scrollOffset = _api.scrollOffset;
 
-    if (_api.invalidatedRows.non_empty())
+    // This if condition serves 2 purposes:
+    // * By setting top/bottom to the full height we ensure that we call Present() without
+    //   any dirty rects and not Present1() on the first frame after the settings change.
+    // * If the scrollOffset is so large that it scrolls the entire viewport, invalidatedRows will span
+    //   the entire viewport as well. We need to set scrollOffset to 0 then, not just because scrolling
+    //   the contents of the entire swap chain is redundant, but more importantly because the scroll rect
+    //   is the subset of the contents that are being scrolled into. If you scroll the entire viewport
+    //   then the scroll rect is empty, which Present1() will loudly complain about.
+    if (_api.invalidatedRows == range<u16>{ 0, _p.s->cellCount.y })
     {
-        const auto deltaPx = _api.scrollOffset * _p.s->font->cellSize.y;
-        const til::CoordType targetSizeX = _p.s->targetSize.x;
-        const til::CoordType targetSizeY = _p.s->targetSize.y;
-        u16 y = 0;
+        _p.MarkAllAsDirty();
+    }
 
-        _p.dirtyRectInPx.left = 0;
-        _p.dirtyRectInPx.top = _api.invalidatedRows.start * _p.s->font->cellSize.y;
-        _p.dirtyRectInPx.right = targetSizeX;
-        _p.dirtyRectInPx.bottom = _api.invalidatedRows.end * _p.s->font->cellSize.y;
+    if (const auto offset = _api.scrollOffset)
+    {
+        if (offset < 0)
+        {
+            // scrollOffset/offset = -1
+            // +----------+    +----------+
+            // |          |    | xxxxxxxxx|
+            // | xxxxxxxxx| -> |xxxxxxx   |
+            // |xxxxxxx   |    |          |
+            // +----------+    +----------+
+            const auto dst = std::copy_n(_p.rows.begin() - offset, _p.rows.size() + offset, _p.rowsScratch.begin());

```

Comment:
I sometimes get write access violation here if `-offset` is greater than `_p.rows.size()`.

![Just before the crash](https://user-images.githubusercontent.com/12915102/236096418-8e23e01b-eb77-48a6-8fa5-d7804475922e.png)
![Access violation in memcpy](https://user-images.githubusercontent.com/12915102/236096632-bd33c2bb-15c9-4d99-bcba-054de5071729.png)

Just wanted to check if https://github.com/microsoft/terminal/issues/14082#issuecomment-1533946883 is fixed by this PR, but running `oh-my-posh debug` a few times crashes Terminal with these changes.

Reply:
Argh, dang, I forgot to write `offset` into `_api.scrollOffset` after clamping it to a valid value range. Thanks for catching this!